Homework 4: Sentiment Analysis - Task 4
----

Names & Sections
----
Names: Julia Geller (4120) and Shae Marks (4120)

Task 4: Neural Networks (20 points)
----

Next, we'll train a feedforward neural net to work with this data. You'll train one neural net which takes the same input as your Logistic Regression model - a sparse vector representing documents as bags of words.

In [1]:
import sentiment_utils as sutils
import numpy as np

from keras.models import Sequential
from keras.layers import Dense

# you can experiment with having some Dropout layers if you'd like to
# this is not required
from keras.layers import Dropout

# if you want to use this again
from sklearn.feature_extraction.text import CountVectorizer



[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\J-Dog\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [2]:
# define constants for the files we are using
TRAIN_FILE = "movie_reviews_train.txt"
DEV_FILE = "movie_reviews_dev.txt"

# load in your data and make sure you understand the format
# Do not print out too much so as to impede readability of your notebook
train_tups = sutils.generate_tuples_from_file(TRAIN_FILE)
dev_tups = sutils.generate_tuples_from_file(DEV_FILE)

# you may use either your sparse vectors or sklearn's CountVectorizer's sparse vectors
# you will experiment with multinomial and binarized representations later


In [3]:
# split the train and devlopment tuples into X and y
X_train = train_tups[0]
y_train = train_tups[1]

X_dev = dev_tups[0]
y_dev = dev_tups[1]

In [4]:
# un-tokenize reviews
X_train_flat = [' '.join(row) for row in X_train]
X_dev_flat = [' '.join(row) for row in X_dev]
# get binary features using Sk-learns Count Vectorizer
X_train_CV_binary, X_dev_CV_binary = sutils.featurize('CV', None, X_train_flat, X_dev_flat, binary = True)

In [5]:
# get multinomial features using Sk-learns Count Vectorizer
X_train_multi, X_dev_CV_multi = sutils.featurize('CV', None, X_train_flat, X_dev_flat, binary = True)

In [6]:
# Create a feedforward neural network model
# that takes a sparse BoW representation of the data as input
# and makes a binary classification of positive/negative sentiment as output
# you may use any number of hidden layers >= 1 and any number of units in each hidden layer (we recommend between 50-200)
# you may use any activation function on the hidden layers 
# you should use a sigmoid activation function on the output layer
# you should use binary cross-entropy as your loss function
# sgd is an appropriate optimizer for this task
# you should report accuracy as your metric
# you may add Dropout layers if you'd like to

# create/compile your model in this cell

hidden_units = 100
num_epochs = 10
input_dim = len(X_train_CV_binary[0])

model = Sequential()

# hidden layer 
model.add(Dense(units=hidden_units, activation='relu', input_dim=input_dim))
# output layer
model.add(Dense(units=1, activation='sigmoid'))

# configure the learning process
model.compile(loss='binary_crossentropy',
              optimizer='sgd',
              metrics=['accuracy'])

model.summary()


Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 100)               2259700   
                                                                 
 dense_1 (Dense)             (None, 1)                 101       
                                                                 
Total params: 2259801 (8.62 MB)
Trainable params: 2259801 (8.62 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


How many trainable parameters does your model have? 2,259,801

In [7]:
# train your model
# Felix's computer takes about 2 sec for 3 epochs
# reports an accuracy of 0.78 at that point using the sgd optimizer

# Failed to find data adapter that can handle input: <class 'numpy.ndarray'>, (<class 'list'> containing values of types {"<class 'int'>"})
# indicates you should change a list into a numpy array

model.fit(X_train_CV_binary, y_train, epochs=num_epochs, verbose=1)


Epoch 1/10
50/50 [==============================] - 2s 19ms/step - loss: 0.6894 - accuracy: 0.5450
Epoch 2/10
50/50 [==============================] - 1s 19ms/step - loss: 0.6675 - accuracy: 0.6956
Epoch 3/10
50/50 [==============================] - 1s 23ms/step - loss: 0.6418 - accuracy: 0.7769
Epoch 4/10
50/50 [==============================] - 1s 20ms/step - loss: 0.6131 - accuracy: 0.8231
Epoch 5/10
50/50 [==============================] - 1s 19ms/step - loss: 0.5831 - accuracy: 0.8494
Epoch 6/10
50/50 [==============================] - 1s 20ms/step - loss: 0.5536 - accuracy: 0.8644
Epoch 7/10
50/50 [==============================] - 1s 21ms/step - loss: 0.5240 - accuracy: 0.8725
Epoch 8/10
50/50 [==============================] - 1s 19ms/step - loss: 0.4952 - accuracy: 0.8813
Epoch 9/10
50/50 [==============================] - 1s 19ms/step - loss: 0.4676 - accuracy: 0.8856
Epoch 10/10
50/50 [==============================] - 1s 21ms/step - loss: 0.4413 - accuracy: 0.8919


In [8]:
# make a prediction on the dev set
# then make a classification decision based on that prediction
# predicting all examples takes < 1 sec on Felix's computer
preds = model.predict(X_dev_CV_binary)
preds = [1 if y >= 0.5 else 0 for y in preds]

7/7 [==============================] - 0s 11ms/step


In [9]:
# use the model.evaluate function to report the loss and accuracy on the dev set
loss, acc = model.evaluate(X_dev_CV_binary, y_dev)
print('loss:', loss, 'accuracy:', acc)

7/7 [==============================] - 1s 10ms/step - loss: 0.5166 - accuracy: 0.7900
loss: 0.5166181325912476 accuracy: 0.7900000214576721


In [11]:
# create the same graph as with NB and LR, with your neural network model instead!
# make sure to re-create your model each time you train it — you don't want to start with
# an already trained network!

# For a model with one hidden layer of 50 units:
# Takes < 15 sec to run on Felix's computer w/ 3 epochs
# Takes < 30 sec to run on Felix's computer w/ 10 epochs
# Takes < 50 sec to run on Felix's computer w/ 20 epochs
# you need not train your model more than 20 epochs
# you should experiment with different numbers of epochs to see how performance varies
# you need not create an experiment that takes > 10 min to run (please do not do this)

# graph for Binary, Count Vectorizer, 5 epochs
train_feats = [(x, y) for x,y in zip(X_train_CV_binary, y_train) ]
dev_feats = [(x, y) for x,y in zip(X_dev_CV_binary, y_dev) ]
sutils.create_training_graph(sutils.neural_net_metrics, train_feats, dev_feats, 'Neural Network', 'Neural_Network_CV_Binary_5Epochs_Graph.png', True, 5)

0.1 of training data used
Epoch 1/5
5/5 [==============================] - 0s 11ms/step - loss: 0.6890 - accuracy: 0.5688
Epoch 2/5
5/5 [==============================] - 0s 10ms/step - loss: 0.6780 - accuracy: 0.6687
Epoch 3/5
5/5 [==============================] - 0s 11ms/step - loss: 0.6664 - accuracy: 0.7063
Epoch 4/5
5/5 [==============================] - 0s 11ms/step - loss: 0.6558 - accuracy: 0.7625
Epoch 5/5
7/7 [==============================] - 0s 4ms/step
0.6461538461538462 0.4 0.4941176470588235 0.57
0.2 of training data used
Epoch 1/5
10/10 [==============================] - 0s 10ms/step - loss: 0.6938 - accuracy: 0.5000
Epoch 2/5
10/10 [==============================] - 0s 10ms/step - loss: 0.6834 - accuracy: 0.5500
Epoch 3/5
10/10 [==============================] - 0s 11ms/step - loss: 0.6740 - accuracy: 0.5750
Epoch 4/5
10/10 [==============================] - 0s 10ms/step - loss: 0.6637 - accuracy: 0.6438
Epoch 5/5
7/7 [==============================] - 0s 5ms/step
0.6

KeyboardInterrupt: 

In [ ]:

# graph for Binary, Count Vectorizer, 10 epochs
train_feats = [(x, y) for x,y in zip(X_train_CV_binary, y_train) ]
dev_feats = [(x, y) for x,y in zip(X_dev_CV_binary, y_dev) ]
sutils.create_training_graph(sutils.neural_net_metrics, train_feats, dev_feats, 'Neural Network', 'Neural_Network_CV_Binary_10Epochs_Graph.png', True, 10)

In [ ]:

# graph for Binary, Count Vectorizer, 10 epochs
train_feats = [(x, y) for x,y in zip(X_train_CV_binary, y_train) ]
dev_feats = [(x, y) for x,y in zip(X_dev_CV_binary, y_dev) ]
sutils.create_training_graph(sutils.neural_net_metrics, train_feats, dev_feats, 'Neural Network', 'Neural_Network_CV_Binary_20Epochs_Graph.png', True, 20)


In [ ]:
# get f1 score for Count Vectorizer, binary, 10 epochs
sutils.neural_net_metrics(X_train_CV_binary, y_train, X_dev_CV_binary, y_dev, 10)[2]

In [ ]:
# get f1 score for Count Vectorizer, multinomial, 10 epochs
sutils.neural_net_metrics(X_train_CV_binary, y_train, X_dev_CV_binary, y_dev, 10)[2]

Report the f1 scores for your model with the following settings, using the same number of epochs to train in both cases:
- number of epochs used: 10
- multinomial features: __YOUR ANSWER HERE__ 
- binarized features: __YOUR ANSWER HERE__